cpp/tests/search/search_list_test.cpp

```diff
@@ -399,8 +409,262 @@ TEST_F(ListLowerBound, ListWithNulls)
+  auto const expected_upper_bound = int32s_col{1, 4, 4, 0, 0, 0, 4, 4};
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_lower_bound, *result_lower_bound, verbosity);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_upper_bound, *result_upper_bound, verbosity);
+}
+
+TEST_F(ListBinarySearch, ListsOfEqualStructsInTwoTables)
+{
+  // Haystack must be pre-sorted.
+  auto const haystack = [] {
+    auto offsets = int32s_col{0, 2, 3, 4, 5, 7, 10, 13, 16, 18};
+    // clang-format off
+    auto data1 = int32s_col{1, 2,
+                            3,
+                            3,
+                            3,
+                            4, 5,
+                            4, 5, 4,
+                            4, 5, 4,
+                            4, 5, 4,
+                            4, 6
+    };
+    auto data2 = int32s_col{1, 2,
+                            3,
+                            3,
+                            3,
+                            4, 5,
+                            4, 5, 4,
+                            4, 5, 4,
+                            4, 5, 4,
+                            5, 1
+    };
+    // clang-format on
+    auto child = structs_col{{data1, data2}};
+    return cudf::make_lists_column(9, offsets.release(), child.release(), 0, {});
+  }();
+
+  auto const needles = [] {
+    auto offsets = int32s_col{0, 2, 3, 4, 5, 7, 10, 13, 15, 17};
+    // clang-format off
+    auto data1 = int32s_col{1, 2,
+                            3,
+                            4,
+                            5,
+                            4, 5,
+                            5, 5, 4,
+                            4, 5, 4,
+                            4, 4,
+                            4, 6
+    };
+    auto data2 = int32s_col{1, 2,
+                            3,
+                            4,
+                            5,
+                            4, 5,
+                            5, 5, 4,
+                            4, 5, 4,
+                            4, 4,
+                            5, 1
+    };
+    // clang-format on
+    auto child = structs_col{{data1, data2}};
+    return cudf::make_lists_column(9, offsets.release(), child.release(), 0, {});
+  }();
+
+  // In this search, the two table have many equal structs.
+  // This help to verify the internal implementation of two-table lex comparator in which the
+  // structs column of two input tables are concatenated, ranked, then split.
+  auto const [result_lower_bound, result_upper_bound] = search_bounds(
+    cudf::table_view{{*haystack}}, cudf::table_view{{*needles}}, {cudf::order::ASCENDING});
+  auto const expected_lower_bound = int32s_col{0, 1, 4, 9, 4, 9, 5, 4, 8};
+  auto const expected_upper_bound = int32s_col{1, 4, 4, 9, 5, 9, 8, 4, 9};
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_lower_bound, *result_lower_bound, verbosity);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_upper_bound, *result_upper_bound, verbosity);
+}
+
+TEST_F(ListBinarySearch, CrazyListTest)
```

Comment:
Love the name. Thank you for writing this test! It gives me a lot more confidence that we're handling the arbitrary depth correctly.